### ISSUE TITLE
BPMN Compensation Events + Activities

### ISSUE THREAD

**Issue by bastianeicher** [camunda/zeebe#5134]
BPMN 2.0 specifies "Compensation Boundary Events" and "Compensation Activities". These provide a handy way to encode "undo in case of failure" logic in workflows.

The [bpmn.io](https://bpmn.io/) online editor supports these constructs, but they are not available in Zeebe Modeler. I assume this means Zeebe does not support them (yet). Is Zeebe support Compensation Activities planned?

**Comment by npepinpe** [comment]
Compensation (activity or boundary event) is not yet planned for Zeebe in the near future. It may be implemented down the line, but that will (most likely) take quite a while. That said, like all feature requests, we'll bring it up with the product team for the next round of planning.

**Comment by saig0** [comment]
@janek64 no updates yet. Compensation events are definitely on our roadmap for new BPMN elements. 

But for the next release, we have other big features on the roadmap like starting a process instance anywhere and process instance modification, etc.

**Comment by saig0** [comment]
@lzgabel I'm happy about your interest. :+1: 

The Zeebe team plan to work on compensation events for the next release `8.2`. External contributions could interfere with our work. 

Please have a look at other issue with a `hacktoberfest` label: https://github.com/camunda/zeebe/issues?q=is%3Aissue+is%3Aopen+label%3Ahacktoberfest

**Comment by amilos** [comment]
@saig0 It seems that compensation events did not make it to 8.2 release which is unfortunate for us as we were looking to move to move a lot of dev teams C7 from to C8. 
- What are the chances of this feature delivered in 8.3? 
- Are you still closed for contributions?

**Comment by remcowesterhoud** [comment]
Thanks for your interest in this topic @amilos! I will discuss this with the team on Monday and get back to you after.

**Comment by abbasadel** [comment]
Hi @amilos, thanks again for your interest in the "compensation" topic. 

Unfortunately, this feature is still closed for contribution since it requires internal coordination. However, it got prioritized for 8.3 roadmap. We will update the issue once we start working on it.

**Comment by amilos** [comment]
Hi @abbasadel, is compensation still prioritized for upcoming 8.3 release or it got deprioritized in the meantime?

**Comment by GerrieSchenck** [comment]
Any updates on this? Will it be included in the 8.3 release?

**Comment by remcowesterhoud** [comment]
Hi @GerrieSchenck and @amilos, thanks for reaching out. Unfortunately I'll have to disappoint you, as compensation will not be part of the 8.3 release. Our focus on other topics left us with no capacity to work on this.

We have no definitive planning for 8.4 and 8.5 yet so I can't give you an indication when this is coming at this time. I will notify our PM so he's aware of your interest in this topic.

/cc @aleksander-dytko fyi 🙂

**Comment by dddpaul** [comment]
Hi.

I have an issue with Camunda 8 and long running processes. By long running process I mean:
- process has an subprocess
- subprocess has an overall timer boundary event set to 10-30 min duration

For example:

<img width="743" alt="image" src="https://github.com/camunda/issues/assets/628226/5757184f-5674-40ee-89b6-602f6f0b4e9c">

With hundreds of thousands PIs being deployed there are plenty of catch timer events could be fired in small time period. With 1 million PI being deployed it may look like this:

<img width="639" alt="image" src="https://github.com/camunda/issues/assets/628226/c6989712-ae4a-42fa-adb4-6a28939f6847">

And it's make me frustrated. I've just want to cancel execution. Instead stream processors are flooded by internal generated events and my brokers are killed by OOM.

The batch processing could help me with this, but to be combined in batch - commands should be generated from one PI, not from multiple ones.

The backpressure mechanism also could be useful - but it limits only commands from client, not internal ones.

There is some feature https://github.com/camunda/zeebe/issues/12560 to present internal stream processor reate limiter, but it's blocked by some mystical https://github.com/camunda/product-hub/issues/1426.

These separate and special compensation events presented by this feature, will they have a separate and special handling or a more controllable handling to prevent event flooding?

PS: I've tested on 8.3.0-alpha6.
PPS: Sorry for duplicating my question across #5134, #12560 and https://github.com/camunda/issues/issues/56, I really need an answers. Thanks in advance!

**Comment by abbasadel** [comment]
Hi @dddpaul ,

Thanks for reporting the issue. 
Did you try to [submit a bug report ](https://github.com/camunda/zeebe/issues/new/choose)? This would be a better way to get fast response rather than duplicating your question across issues. You can also checkout Camunda's [community forum](https://forum.camunda.io/?utm_medium=event&utm_source=bevy), amazing developers like you are helping each other over there.

**Comment by remcowesterhoud** [comment]
Please don't raise another issue. We already have it listed [here](https://github.com/camunda/zeebe/issues/13870). You can subscribe to this issue if you'd like to follow the progress.

In the meantime you can enable the yielding timer due date checker (`ZEEBE_BROKER_EXPERIMENTAL_FEATURES_ENABLEYIELDINGDUEDATECHECKER`). This is enabled by default as of version 8.3. This will make it so the triggering of timers yields to unblock other processing.

**Comment by saig0** [comment]
> These separate and special compensation events presented by this feature, will they have a separate and special handling or a more controllable handling to prevent event flooding?

Compensation events will be different. Compared to timer events, compensation is triggered actively within the process using a compensation throw event (or similar). It is unlikely that many compensation events will be triggered at the "same time". 

I don't expect that we will face similar issues as for timers. However, when we did implement compensation, please go ahead and do any performance testing to discover bottlenecks. :rocket:

**Comment by saig0** [comment]
Closing this issue in favor of the new epic issue for adding support for compensation events: https://github.com/camunda/zeebe/issues/14920.
